### ISSUE TITLE
How to limit the allowed actions based on the current state for the agent?

### ISSUE THREAD

**Issue by aheidariiiiii1993** [DLR-RM/stable-baselines3#1352]
### ❓ Question

Hello,
I am using QR-DQN agent of sb3_contrib. In my problem, I have 9 discrete actions but for some specific states a set of actions are obviously wrong. So I want to write some if then rules that tells to the agent in that specific state, he can only choose from a limited set of actions. Is it possible to be done in sb3? If there is any example code I would appreciate it. 
Thank you 

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] If code there is, it is minimal and working
- [X] If code there is, it is formatted using the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by araffin** [comment]
Hello,

> Is it possible to be done in sb3? If there is any example code I would appreciate it.

You have maskable ppo in SB3 contrib that is made for that.
For QR-DQN, you will need to adapt the code to use the mask passed by the env (please check the implementation of maskable ppo) to reduce the q-values of the invalid actions (so they are not selected).

You should also be aware of that: https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/issues/145
(you will need to override `action_space.sample()` for the warmup phase.

related: https://github.com/DLR-RM/stable-baselines3/issues/269 https://github.com/DLR-RM/stable-baselines3/issues/580

**Comment by aheidariiiiii1993** [comment]
Thank you @araffin for your reply. 

I can actually use maskable ppo in my problem. However, I coult not yet understand how should I use it to use my constraints. I only found [the minimal example](https://sb3-contrib.readthedocs.io/en/master/modules/ppo_mask.html) and [this code](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/blob/master/sb3_contrib/common/envs/invalid_actions_env.py#L35-L36) as available instructions though it is not yet clear for me how to code my constaints. 

In my case, I have a **custom environment** with 9 discrete actions. And I want to set the constraint that when the environment.T_indoor <20 then the actions 3 and 4 should be impossible. 

Is there any example that I can use to learn how to program maskable ppo with my constraints?
Thank you

**Comment by aheidariiiiii1993** [comment]
Thank you very much @araffin for your guidance. Now I am using maskable ppo to see if I can get better results. 

However, I am wondering that when I am training the model, it seems multiple episodes are running in parallel (or maybe multiple environments) because there are multiple green bars at the same time. It is (1) extremely slow in my case since my **custom environment** is calling an external software, and (2) at the end total number of timesteps that the model was trained was more than what I set. 

Is there anyway that I deactivate it, and the training process would be just like other agents (one episode after the other)? Here is my code

````
def mask_fn(environment: gym.Env):
    
    mask = np.ones(environment.action_space.n, dtype=np.int32)
    
    if (environment.T_avg <20 & environment.occ ==1):
        
        mask[[1, 3, 5, 7, 8]] = 0
        
    if (environment.T_avg>24 & environment.occ ==1):
        
        mask[[0,2,4,5,8]] = 0 
       
    return mask 

environment = ActionMasker(environment, mask_fn)  

model = MaskablePPO(MaskableActorCriticPolicy, 
                     environment,
                     verbose=0,
                     learning_rate=0.0005,  
                     n_steps=48,
                     batch_size=48,  
                     gamma=0.9,  
                     policy_kwargs=None,
                     device='auto')

for iteration in range (0,number_of_iterations):
    
    environment.sim_start_time_actual=1416 # what is actually in practice
    
    environment.sim_stop_time_actual=1416+1

    environment.sim_start_time_trnsys=1416

    environment.sim_stop_time_trnsys=1416+1

    model.learn(total_timesteps = total_timesteps,
                log_interval = 2,
                reset_num_timesteps = True,
                progress_bar= True)
    
    model.save(saving_folder_agent + "\model_checkpoint_{}".format(iteration))
````

**Comment by aheidariiiiii1993** [comment]
Hi @araffin , thank you for your reply. 
I am now using maskable ppo. However, it seems that the use of environment in the training process is quite different from other agents. In my case I have a custom environment that calls an external software. The environment sets the timestep of external software as 1416 and runs it for one timestep in each step of environment. Using the other agents it is all fine, but sing the maskable ppo I get a weired behavior (for example the timesteps anyway starts from 0 and instead of runing it for one month 12 times, it is runing it for 12 months). I tested the code with QRDQN agent and it is working well. So in maskable ppo that the interation with the environment should be different from all the other agents (parrallel environments or something). Is there anyway to deactivate it so it would be like the other agents?


```

def mask_fn(environment: gym.Env):
    
    mask = np.ones(environment.action_space.n, dtype=np.int32)
    
    if environment.occ == 1: 
    
        if environment.T_avg <20:

            mask[[1, 3, 5, 7, 8]] = 0

        if environment.T_avg>24:

            mask[[0, 2, 4, 6, 8]] = 0 
       
    return mask 


environment = ActionMasker(environment, mask_fn)  

model = MaskablePPO(MaskableActorCriticPolicy, 
                     environment,
                     verbose=0,
                     learning_rate=0.0005,  
                     n_steps=48,
                     batch_size=48,  
                     gamma=0.9,  
                     device='auto')


for iteration in range (0,number_of_iterations):
    
    environment.sim_start_time_actual=1416 # what is actually in practice
    
    environment.sim_stop_time_actual=1416+1

    environment.sim_start_time_trnsys=1416

    environment.sim_stop_time_trnsys=1416+1

    model.learn(total_timesteps = total_timesteps,
                log_interval = 2,
                reset_num_timesteps = True,
                progress_bar= True)
    
    model.save(saving_folder_agent + "\model_checkpoint_{}".format(iteration))
```

**Comment by aheidariiiiii1993** [comment]
Hi @araffin , thank you for your reply.
I am now using maskable ppo. However, it seems that the use of environment in the training process is quite different from other agents. In my case I have a custom environment that calls an external software. The environment runs the external software and gives back the new state and reward. I used my code with QRDQN agent and it is working well. But in maskable ppo it seems that the interation with the environment is different from all the other agents (it multiple environments?!). Is there anyway to deactivate it so it would be like the other agents? Once my environment calls an external software it does not make sense to have multipl environments. I looked at the source code of maskable ppo and played with it a bit but I could not understand how to use only one single environment.

```
def mask_fn(environment: gym.Env):
    
    mask = np.ones(environment.action_space.n, dtype=np.int32)
    
    if environment.occ == 1: 
    
        if environment.T_avg <20:

            mask[[1, 3, 5, 7, 8]] = 0

        if environment.T_avg>24:

            mask[[0, 2, 4, 6, 8]] = 0 
       
    return mask 


environment = ActionMasker(environment, mask_fn)  

model = MaskablePPO(MaskableActorCriticPolicy, 
                     environment,
                     verbose=0,
                     learning_rate=0.0005,  
                     n_steps=48,
                     batch_size=48,  
                     gamma=0.9,  
                     device='auto')


for iteration in range (0,number_of_iterations):
    
    environment.sim_start_time_actual=1416 # what is actually in practice
    
    environment.sim_stop_time_actual=1416+1

    environment.sim_start_time_trnsys=1416

    environment.sim_stop_time_trnsys=1416+1

    model.learn(total_timesteps = total_timesteps,
                log_interval = 2,
                reset_num_timesteps = True,
                progress_bar= True)
    
    model.save(saving_folder_agent + "\model_checkpoint_{}".format(iteration))
```
